vhost/src/vhost_user/gpu_backend_req.rs

```diff
@@ -0,0 +1,133 @@
+use crate::vhost_user;
+use crate::vhost_user::connection::Endpoint;
+use crate::vhost_user::gpu_message::{
+    GpuBackendReq, VhostUserGpuMsgHeader, VirtioGpuRespDisplayInfo,
+};
+use crate::vhost_user::message::VhostUserMsgValidator;
+use crate::vhost_user::Error;
+use std::os::fd::RawFd;
+use std::os::unix::net::UnixStream;
+use std::sync::{Arc, Mutex, MutexGuard};
+use std::{io, mem};
+use vm_memory::ByteValued;
+
+use super::gpu_message::{VirtioGpuGetEdid, VirtioGpuRespGetEdid};
+
+struct BackendInternal {
+    sock: Endpoint<VhostUserGpuMsgHeader<GpuBackendReq>>,
+    // whether the endpoint has encountered any failure
+    error: Option<i32>,
+}
+
+impl BackendInternal {
+    fn check_state(&self) -> vhost_user::Result<u64> {
+        match self.error {
+            Some(e) => Err(Error::SocketBroken(io::Error::from_raw_os_error(e))),
+            None => Ok(0),
+        }
+    }
+
+    fn send_header<V: ByteValued + Sized + Default + VhostUserMsgValidator>(
+        &mut self,
+        request: GpuBackendReq,
+        fds: Option<&[RawFd]>,
+    ) -> vhost_user::Result<V> {
+        self.check_state()?;
+
+        let hdr = VhostUserGpuMsgHeader::new(request, 0, 0);
+        self.sock.send_header(&hdr, fds)?;
+
+        self.wait_for_ack(&hdr)
+    }
+
+    fn send_message<T: ByteValued, V: ByteValued + Sized + Default + VhostUserMsgValidator>(
+        &mut self,
+        request: GpuBackendReq,
+        body: &T,
+        fds: Option<&[RawFd]>,
+    ) -> vhost_user::Result<V> {
+        self.check_state()?;
+
+        //investigate what len should be
+        let len = mem::size_of::<T>();
+        let hdr = VhostUserGpuMsgHeader::new(request, 0, len as u32);
+        self.sock.send_message(&hdr, body, fds)?;
+
+        self.wait_for_ack(&hdr)
+    }
+
+    fn wait_for_ack<V: ByteValued + Sized + Default + VhostUserMsgValidator>(
+        &mut self,
+        hdr: &VhostUserGpuMsgHeader<GpuBackendReq>,
+    ) -> vhost_user::Result<V> {
+        self.check_state()?;
+        let (reply, body, rfds) = self.sock.recv_body::<V>()?;
```

Comment:
Hello @mtjhrc , do you mind to take  a look at this function? in `recv_body()`  at https://github.com/mtjhrc/vhost/blob/89126226b5df523bbeaffc935ba3cce816d82139/vhost/src/vhost_user/connection.rs#L499 it error partialmessage for getEdid request.